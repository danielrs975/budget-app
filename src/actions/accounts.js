/**
 * This script contain all the actions about the
 * accounts information
 * @author Daniel Rodriguez
 */
import uuid from 'uuid';

/**
 * This action add an account to the database
 * @param {*} account all the info about the account
 */
export const addAccount = (account) => ({
	type    : 'ADD_ACCOUNT',
	account
});

export const startAddAccount = (account) => {
	return (dispatch, getState) => {
		dispatch(addAccount({ id: uuid(), ...account }));
		const accounts = getState().accounts;
		localStorage.setItem('accounts', JSON.stringify(accounts));
	};
};

/**
 * This action is set in the state the accounts that are saved
 * in the db
 * @param {*} accounts The accounts to set in the state
 */
export const setAccount = (accounts) => ({
	type     : 'SET_ACCOUNTS',
	accounts
});

export const startSetAccount = () => {
	return (dispatch) => {
		const accounts = JSON.parse(localStorage.getItem('accounts'));
		dispatch(setAccount(accounts || []));
	};
};
