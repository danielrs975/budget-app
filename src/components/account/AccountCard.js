/**
 * This component contain a short information about an account
 * @author Daniel Rodriguez
 */
import React from 'react';
import { Link } from 'react-router-dom';

const AccountCard = ({ name, amount, id }) => (
	<div className="card-layout">
		<Link to={`account/${id}`}><h3>{name}</h3></Link>
		<p>{amount / 100}</p>
	</div>
);

export default AccountCard;
