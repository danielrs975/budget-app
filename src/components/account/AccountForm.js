import React from 'react';

export default class AccountForm extends React.Component {
	constructor(props) {
		super(props);
		const { account } = props;
		this.state = {
			name     :
				account ? account.name :
				'',
			amount   :
				account ? account.amount :
				'',
			type     :
				account ? account.type :
				'bank',
			currency :
				account ? account.currency :
				'euro',
			error    : undefined
		};
	}
	onNameChange = (e) => {
		const name = e.target.value;
		this.setState(() => ({ name }));
	};
	onAmountChange = (e) => {
		const amount = e.target.value;
		if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
			this.setState(() => ({ amount }));
		}
	};
	onTypeChange = (e) => {
		const type = e.target.value;
		this.setState(() => ({ type }));
	};
	onCurrencyChange = (e) => {
		const currency = e.target.value;
		this.setState(() => ({ currency }));
	};
	saveAccount = (e) => {
		e.preventDefault();
		if (!this.state.name || !this.state.currency) {
			this.setState(() => ({ error: 'A name for the account must be provided' }));
		} else {
			this.setState(() => ({ error: undefined }));
			this.props.onSubmit({
				name     : this.state.name,
				amount   : parseFloat(this.state.amount) * 100,
				type     : this.state.type,
				currency : this.state.currency
			});
		}
	};
	render() {
		return (
			<div>
				<div className="content-container">
					<form onSubmit={this.saveAccount}>
						{this.state.error && <p>{this.state.error}</p>}
						<input
							type="text"
							placeholder="Name of account"
							value={this.state.name}
							onChange={this.onNameChange}
						/>
						<input
							type="text"
							placeholder="Initial amount"
							value={this.state.amount}
							onChange={this.onAmountChange}
						/>
						<select placeholder="Type" value={this.state.type} onChange={this.onTypeChange}>
							<option value="bank">Bank</option>
							<option value="cash">Cash</option>
						</select>
						<select placeholder="Currency" value={this.state.currency} onChange={this.onCurrencyChange}>
							<option value="euro">Euro</option>
							<option value="dollar">Dollar</option>
						</select>
						<button>Save account</button>
					</form>
				</div>
			</div>
		);
	}
}
