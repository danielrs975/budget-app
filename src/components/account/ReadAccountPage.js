/**
 * Component that show the details of an
 * account
 * @author Daniel Rodriguez
 */
import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import TransactionsTable from '../transaction/TransactionsTable';

export const ReadAccountPage = ({ account }) => {
    return (
        <div>
            <div className="page-header">
                <div className="content-container">
                    <h1>{account.name}</h1>
                </div>
            </div>
            <div className="content-container">
                <p>Remaining Amount: {account.amount / 100}</p>
                <Link to="/transaction/create">Add a Transaction</Link>
            </div>
            <TransactionsTable />
        </div>
    );
}

const mapStateToProps = (state, props) => ({
    account: state.accounts.find((account) => account.id === props.match.params.id)
});

export default connect(mapStateToProps)(ReadAccountPage);