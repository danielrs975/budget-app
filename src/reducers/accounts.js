/**
 * This is the reducer for the accounts
 * information
 * @author Daniel Rodriguez
 */

export default (state = [], action) => {
	switch (action.type) {
		case 'ADD_ACCOUNT':
			return [ ...state, action.account ];
		default:
			return state;
	}
};
