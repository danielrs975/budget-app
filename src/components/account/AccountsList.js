import React from 'react';
import AccountCard from './AccountCard';
import { Link } from 'react-router-dom';

const AccountsList = ({ accounts }) => (
	<div>
		{accounts.length > 0 &&
			accounts.map((account) => <AccountCard key={account.id} {...account} />)}
		<Link to="account/create">Add Account</Link>
	</div>
);

export default AccountsList;
