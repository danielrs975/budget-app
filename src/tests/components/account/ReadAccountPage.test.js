import React from 'react';
import { shallow } from 'enzyme';
import { ReadAccountPage } from '../../../components/account/ReadAccountPage';
import accounts from '../../fixtures/accounts';

test('should render ReadAccountPage correctly', () => {
    const wrapper = shallow(<ReadAccountPage account={accounts[0]} />)
    expect(wrapper).toMatchSnapshot();
})