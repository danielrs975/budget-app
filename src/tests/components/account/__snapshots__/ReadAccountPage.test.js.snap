// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ReadAccountPage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1>
        Revolut
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <p>
      Remaining Amount: 
      10
    </p>
    <[object Object]
      to="/transaction/create"
    >
      Add a Transaction
    </[object Object]>
  </div>
</div>
`;
